src/main/java/com/atlassian/db/replica/internal/ReplicaStatement.java

```diff
@@ -558,37 +559,22 @@ public Statement getReadStatement(RouteDecisionBuilder decisionBuilder) {
             connectionProvider.getStateDecision().ifPresent(decisionBuilder::cause);
             return prepareWriteStatement(decisionBuilder);
         }
-        final String sql = decisionBuilder.getSql();
-        isWriteOperation = sqlFunction.isFunctionCall(sql) || isUpdate(sql) || isDelete(sql);
-        if (isWriteOperation) {
+        SqlQuery sqlQuery = new SqlQuery(decisionBuilder.getSql());
+        if (sqlQuery.isWriteOperation(sqlFunction)) {
             decisionBuilder.reason(WRITE_OPERATION);
             return prepareWriteStatement(decisionBuilder);
         }
 
-        if (isSelectForUpdate(sql)) {
+        if (sqlQuery.isSelectForUpdate()) {
```

Comment:
Previous discussion: https://github.com/atlassian-labs/db-replica/pull/49#discussion_r585406595

Reply:
@dagguh
So locking is a type of a write. The SqlQuery.writesToDatabase should catch that. So "writing to DB" is a Reason to go to replica, but there are multiple types of writes. Perhaps SqlQuery needs to return something more than a boolean. Maybe it needs more explitict methods:

boolean mustWriteToDb();
boolean mayWriteToDb();
String getWritingKind(); // e.g. LOCK, UPDATE, DELETE, FUNCTION
 
@wyrzyk wyrzyk 7 days ago Member
What's the use case for revealing these low-level details in the API.

 
@dagguh dagguh 7 days ago Member
I'm talking about internals. We keep hearing that "LOCK" is a type of write. Here's a way to model that explicitly.

 
@wyrzyk wyrzyk 7 days ago Member
Make sense.

 
@afaruga-atlassian afaruga-atlassian yesterday Author Member
Actually LOCK isn't type o write, but it's treat almost as write as it run for on main. acutall. LOCK == SELECT FOR UPDATE
I added isWrite information to Rason on compile level , this way this model say that explicitly. Thanks for this I could remove isWriteOpearatiion internal state, before doing this I've improved code coverage for SELECT FOR UPDATE.

 
@dagguh dagguh 22 hours ago Member
If it's not a write, then why is it run on main?

 
@afaruga-atlassian afaruga-atlassian 21 hours ago Author Member
I guess this is related with the impact on transaction

FOR UPDATE causes the rows retrieved by the SELECT statement to be locked as though for update. This prevents them from being modified or deleted by other transactions until the current transaction ends.
docs

@wyrzyk should provide full answer on this.

 
@wyrzyk wyrzyk 21 hours ago Member
Are other connections affected by the lock? If it's not a write operation, why are they affected?

 
@dagguh dagguh 21 hours ago Member
If we send SELECT FOR UPDATE to replica - can it succeed?

 
@afaruga-atlassian afaruga-atlassian 20 hours ago Author Member
Do we treat connections/calls withing the same connection to Database independently?
I think not, we had this hidden contract that read after write is treated RUN ON MAIN.
I think with SELECT FOR UPDATE the idea was the same.